packages/flutter/lib/src/material/list_tile.dart

```diff
@@ -385,91 +408,563 @@ class ListTile extends StatelessWidget {
+    final EdgeInsets resolvedContentPadding = contentPadding?.resolve(textDirection)
+      ?? tileTheme?.contentPadding?.resolve(textDirection)
+      ?? _kDefaultContentPadding;
+
     return new InkWell(
       onTap: enabled ? onTap : null,
       onLongPress: enabled ? onLongPress : null,
       child: new Semantics(
         selected: selected,
         enabled: enabled,
-        child: new ConstrainedBox(
-          constraints: new BoxConstraints(minHeight: tileHeight),
-          child: new Padding(
-            padding: const EdgeInsets.symmetric(horizontal: 16.0),
-            child: new UnconstrainedBox(
-              constrainedAxis: Axis.horizontal,
-              child: new SafeArea(
-                top: false,
-                bottom: false,
-                child: new Row(children: children),
-              ),
-            ),
-          )
+        child: new SafeArea(
+          top: false,
+          bottom: false,
+          minimum: resolvedContentPadding,
+          child: new _ListTile(
+            leading: leadingIcon,
+            title: titleText,
+            subtitle: subtitleText,
+            trailing: trailingIcon,
+            isDense: _isDenseLayout(tileTheme),
+            isThreeLine: isThreeLine,
+          ),
         ),
       ),
     );
   }
 }
+
+class _ListTile extends RenderObjectWidget {
+  const _ListTile({
+    Key key,
+    this.leading,
+    this.title,
+    this.subtitle,
+    this.trailing,
+    this.isThreeLine,
+    this.isDense,
+  }) : super(key: key);
+
+  final Widget leading;
+  final Widget title;
+  final Widget subtitle;
+  final Widget trailing;
+  final bool isThreeLine;
+  final bool isDense;
+
+  @override
+  _RenderListTileElement createElement() => new _RenderListTileElement(this);
+
+  @override
+  _RenderListTile createRenderObject(BuildContext context) {
+    return new _RenderListTile(
+      isThreeLine: isThreeLine,
+      isDense: isDense,
+      textDirection: Directionality.of(context),
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderListTile renderObject) {
+    renderObject
+      ..isThreeLine = isThreeLine
+      ..isDense = isDense
+      ..textDirection = Directionality.of(context);
+  }
+}
+
+// Identifies the children of a _ListTileElement.
+enum _ListTileSlot {
+  leading,
+  title,
+  subtitle,
+  trailing,
+}
+
+class _RenderListTile extends RenderBox {
+  _RenderListTile({
+    bool isDense,
+    bool isThreeLine,
+    TextDirection textDirection,
+  }) : _isDense = isDense,
+       _isThreeLine = isThreeLine,
+       _textDirection = textDirection;
+
+  static const double _kMinLeadingWidth = 40.0;
+  static const double _kTitleGap = 16.0; // between the titles and the leading/trailing widgets
+
+  final Map<_ListTileSlot, RenderBox> slotToChild = <_ListTileSlot, RenderBox>{};
+  final Map<RenderBox, _ListTileSlot> childToSlot = <RenderBox, _ListTileSlot>{};
```

Comment:
this is a pattern we've seen several times now, we might want to factor it out into a common foundation class